### DISCUSSION TITLE
I can't exclude optional fields when mocking prisma methods

### DISCUSSION THREAD

**Discussion by amirhhashemi** [prisma/prisma#10234] [Q&A]
I exactly followed the instructions on mocking Prisma using `jest-mock-extended`. Let's say this is my db schema:
```prisma
model User {
  id                String   @id @default(cuid())
  name          String   @db.VarChar(100)
  password    String
  bio              String?

  createdAt DateTime @default(now())
}
```
and I have a wrapper method that creates a new user:
```ts
interface Data {
  name: string;
  password: string;
}

export const createUser = async (data: Data) => {
  const newUser = await prisma.user.create({
    data: { ...data },
    select: {id: true, name: true}
  });

  return newUser;
};
```
I should be able to create a new user with just a name and password. this method just returns the id and the name but when I want to mock its resolved value I have to pass other fields too.

```ts
test("should create new user ", async () => {
  const user = {
    name: "Rich",
    password: "1234",
  };

  prismaMock.user.create.mockResolvedValue(user);        // Error: Fields 'id', 'bio' and 'createdAt' are missed
});
```
the same thing applies to other Prisma methods.

I have about 20 columns in the User table and passing all of them every time I want to mock a Prisma function is messy. Why does this happening and is there is way to get around it?

codesandbox: https://codesandbox.io/s/keen-hill-5l4y2?file=/src/create.test.ts

**Comment by austincrim** [comment]
Hi @ahhshm 👋, thanks for the question.

Are you using the example Prisma schema [found here](https://www.prisma.io/docs/guides/testing/unit-testing#example-unit-tests)? If not, could you share your schema with me?

The example schema includes this model:
```prisma
model User {
  id                       Int     @id @default(autoincrement())
  email                    String  @unique
  name                     String?
  acceptTermsAndConditions Boolean
}
```

Which does not include `username` or `password` fields and would explain the type error you are receiving.

Let me know!

**Reply by austincrim** [reply]
Thanks for the Codesandbox!

I think the root issue here is the `prisma.user.create` function returns a full `User` type. So, when you try to mock the return value with `prismaMock.user.create.mockResolvedValue(user)`, the `user` object that gets passed in doesn't have all the required properties to satisfy the `User` type that needs to be returned from the `create` call.

To solve this, you would either need to include all the `User` fields in the `const user` object or mark the excluded fields as optional in your Prisma schema.

Let me know if I can clarify anything further!

**Reply by amirhhashemi** [reply]
Sorry I updated my question a little and deleted my comments so it would be easier for others to follow. I respond to you again so others understand what's going on.

I have to include optional fields too. bio is completely optional.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
